src/coreclr/jit/lclmorph.cpp

```diff
@@ -1013,19 +1026,28 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
                 structLayout = indir->AsBlk()->GetLayout();
             }
 
-            // We're not going to produce a TYP_STRUCT LCL_FLD so we don't need the field sequence.
-            fieldSeq = nullptr;
+            if (fieldSeq != nullptr)
+            {
+                // We're not going to produce a TYP_STRUCT LCL_FLD so we don't need the field sequence.
+                fieldSeq = nullptr;
+                JITDUMP("Dropped field seq from [%06u]\n", m_compiler->dspTreeID(indir));
+            }
         }
 
         // We're only processing TYP_STRUCT variables now so the layout should never be null,
         // otherwise the below layout equality check would be insufficient.
         assert(varDsc->GetLayout() != nullptr);
 
-        if ((val.Offset() == 0) && (structLayout != nullptr) &&
+        if ((val.Offset() == 0) && (structLayout != nullptr) && (fieldSeq == nullptr) &&
             ClassLayout::AreCompatible(structLayout, varDsc->GetLayout()))
         {
             indir->ChangeOper(GT_LCL_VAR);
             indir->AsLclVar()->SetLclNum(val.LclNum());
+
+            if (structLayout->GetClassHandle() != varDsc->GetLayout()->GetClassHandle())
```

Comment:
Can you add a comment here, about the Class handle mismatch, and how/why we handle it